File: src/main/java/moa/moabackend/global/error/exception/ErrorCode.java
package moa.moabackend.global.error.exception;

import lombok.Getter;
import org.springframework.http.HttpStatus;

@Getter
public enum ErrorCode {

    // Common
    INTERNAL_SERVER_ERROR(HttpStatus.INTERNAL_SERVER_ERROR, "서버 내부 오류"),
    INVALID_INPUT_VALUE(HttpStatus.BAD_REQUEST, "잘못된 입력값"),
    METHOD_NOT_ALLOWED(HttpStatus.METHOD_NOT_ALLOWED, "허용되지 않은 HTTP 메서드"),
    ENTITY_NOT_FOUND(HttpStatus.NOT_FOUND, "요청한 엔티티를 찾을 수 없습니다"),
    INVALID_TYPE_VALUE(HttpStatus.BAD_REQUEST, "잘못된 타입의 입력값"),

    // User
    USER_NOT_FOUND(HttpStatus.NOT_FOUND, "사용자를 찾을 수 없습니다"),
    EMAIL_ALREADY_EXISTS(HttpStatus.CONFLICT, "이미 존재하는 이메일입니다"),

    // Auth
    UNAUTHORIZED(HttpStatus.UNAUTHORIZED, "인증이 필요합니다"),
    ACCESS_DENIED(HttpStatus.FORBIDDEN, "권한이 없습니다");

    private final HttpStatus status;
    private final String message;

    ErrorCode(HttpStatus status, String message) {
        this.status = status;
        this.message = message;
    }
}

File: src/main/kotlin/moa/moabackend/global/error/exception/ErrorCode.kt
package moa.moabackend.global.error.exception

import com.fasterxml.jackson.annotation.JsonFormat

@JsonFormat(shape = JsonFormat.Shape.OBJECT)
enum class ErrorCode(
    val status: Int,
    val message: String
) {
    // server error
    INTERNAL_SERVER_ERROR(500, "Internal Server Error"),

    // client error
    BAD_REQUEST(400, "Bad Request"),

    // jwt
    UNAUTHORIZED(401, "Unauthorized"),
    INVALID_TOKEN(401, "Invalid Token"),
    EXPIRED_TOKEN(401, "Expired Token"),

    // user
    USER_NOT_FOUND(404, "User not found"),
    USER_DELETED(403, "User deleted"),
    USER_ALREADY_EXISTS(409, "User already exists"),
    PASSWORD_MISMATCH(401, "Password mismatch"),

    // group purchase
    GROUP_PURCHASE_NOT_FOUND(404, "Group purchase not found"),
    ALREADY_JOINED(409, "Already joined this group purchase"),
    OVER_TARGET_COUNT(400, "Already reached target count"),
    EXPIRED_DEADLINE(400, "Deadline has passed"),
    CANNOT_JOIN_OWN_PURCHASE(400, "Cannot join your own group purchase"),

    // auth
    INVALID_ADMIN_KEY(401, "Invalid admin key"),
    ;
}
